v2/tools/generator/internal/codegen/pipeline/resource_registration_file.go

```diff
@@ -503,3 +470,103 @@ func (r *ResourceRegistrationFile) getImportedPackages() map[astmodel.PackageRef
 
 	return result
 }
+
+func (r *ResourceRegistrationFile) makeWatchesExpr(typeName astmodel.TypeName, codeGenerationContext *astmodel.CodeGenerationContext) dst.Expr {
+	secretWatchesExpr := r.makeSecretWatchesExpr(typeName, codeGenerationContext)
+
+	if secretWatchesExpr == nil {
+		return nil
+	}
+
+	sliceBuilder := astbuilder.NewSliceLiteralBuilder(astmodel.WatchRegistrationType.AsType(codeGenerationContext), true)
+	sliceBuilder.AddElement(secretWatchesExpr)
+	return sliceBuilder.Build()
+}
+
+func (r *ResourceRegistrationFile) makeSecretWatchesExpr(typeName astmodel.TypeName, codeGenerationContext *astmodel.CodeGenerationContext) dst.Expr {
+	secretKeys, ok := r.secretPropertyKeys[typeName]
+	if !ok {
+		return nil
+	}
+
+	if len(secretKeys) == 0 {
+		return nil
+	}
+
+	newWatchBuilder := astbuilder.NewCompositeLiteralBuilder(astmodel.WatchRegistrationType.AsType(codeGenerationContext))
+	sort.Strings(secretKeys)
+
+	// Not using astbuilder here because we want this to go onto a single line
+	source := &dst.CompositeLit{
+		Type: astmodel.ControllerRuntimeSourceKindType.AsType(codeGenerationContext),
+		Elts: []dst.Expr{
+			&dst.KeyValueExpr{
+				Key: dst.NewIdent("Type"),
+				Value: astbuilder.AddrOf(
+					&dst.CompositeLit{
+						Type: astmodel.SecretType.AsType(codeGenerationContext),
+					}),
+			},
+		},
+	}
+	newWatchBuilder.AddField("Src", astbuilder.AddrOf(source))
+
+	secretKeyParams := make([]dst.Expr, 0, len(secretKeys))
+	for _, secretKey := range secretKeys {
+		secretKeyParams = append(secretKeyParams, astbuilder.StringLiteral(secretKey))
+	}
+
+	listTypeName := typeName.WithName(typeName.Name() + "List").AsType(codeGenerationContext)
+
+	eventHandler := astbuilder.CallFunc(
+		"watchSecretsFactory",
+		astbuilder.SliceLiteral(dst.NewIdent("string"), secretKeyParams...),
+		astbuilder.AddrOf(astbuilder.NewCompositeLiteralBuilder(listTypeName).Build()))
+	newWatchBuilder.AddField("MakeEventHandler", eventHandler)
+
+	return newWatchBuilder.Build()
+}
+
+func (r *ResourceRegistrationFile) makeConfigMapWatchesExpr(typeName astmodel.TypeName, codeGenerationContext *astmodel.CodeGenerationContext) dst.Expr {
+	configMapKeys, ok := r.configMapPropertyKeys[typeName]
+	if !ok {
+		return nil
+	}
+
+	if len(configMapKeys) == 0 {
+		return nil
+	}
+
+	newWatchBuilder := astbuilder.NewCompositeLiteralBuilder(astmodel.WatchRegistrationType.AsType(codeGenerationContext))
+	sort.Strings(configMapKeys)
+
+	// Not using astbuilder here because we want this to go onto a single line
```

Comment:
maybe extract this into a method with a parameter `Type` so that can be used by secretExpr as well?

Reply:
I updated the `astbuilder` helper so that we could use it here. It's not tons shorter but it it somewhat shorter.

Reply:
Or did you mean this whole method? I can look into that although there's more to it than just a single type parameter.

Reply:
Reduced duplication between the `secrets` and `configmap` paths here by refactoring to use a `makeSimpleWatchesExpr`.

We might need other variants in the future, but for now the two are similar enough that they can share the same code path.